src/librustc_span/symbol.rs

```diff
@@ -1387,44 +1412,121 @@ impl fmt::Display for MacroRulesNormalizedIdent {
+///
+/// - The highest bit of the u32 is a tag bit.
+///
+/// - If the tag bit is 0, the symbol is inlined. The string bytes
+///   (`s`) are encoded into the u32 (`u`) in a little-endian fashion:
+///   - Byte 0 of `u` holds `s[0]`, or zero if `s.len()` is < 1.
+///   - Byte 1 of `u` holds `s[1]`, or zero if `s.len()` is < 2.
+///   - Byte 2 of `u` holds `s[2]`, or zero if `s.len()` is < 3.
+///   - Byte 3 of `u` holds `s[3]`, or zero if `s.len()` is < 4.
 ///
-/// Note that `Symbol` cannot directly be a `rustc_index::newtype_index!` because it
-/// implements `fmt::Debug`, `Encodable`, and `Decodable` in special ways.
+///   Some examples:
+///   - ""     -> 0x00_00_00_00
+///   - "a"    -> 0x00_00_00_61
+///   - "ab"   -> 0x00_00_62_61
+///   - "abc"  -> 0x00_63_62_61
+///   - "abcd" -> 0x64_63_62_61
+///
+///   Because byte 3 contains the tag bit, for a string of length four to be
+///   inlined its final byte must be < 0x80. For example, "cdé" can not be
+///   inlined because in UTF-8 it is a four byte sequence `[0x64, 0x65, 0xc3,
+///   0xa9]` and byte 3 is > 0x7f. (This is the first of the abovementioned
+///   encoding constraints.)
+///
+///   The length of the string is not explicitly encoded. Rather, it is equal
+///   to the index of the most-significant zero byte, or four if there are no
+///   zero bytes. (For example, in 0x00_63_62_61 the most significant zero byte
+///   is byte 3, and so the length is 3.) Because of this, a string whose final
+///   char is a NUL char cannot be represented inline. For example, none of
+///   "\0", "a\0", "ab\0", and "abc\0" can be inlined. (This is the second of
+///   the abovementioned encoding constraints.)
+///
+/// - If the tag bit is 1, the symbol is tabled. The high bit must be removed
+///   before being used as an index. For example, the fifth tabled symbol will
+///   have the value 0x80_00_00_05.
+///
+///   The maximum value of a `newtype_index` is 0xff_ff_ff_00, due to highest
+///   256 values being reserved. Therefore, the maximum index representable is
+///   0x7f_ff_ff_00 (2,147,483,392). Given that the maximum number of bytes in
+///   a crate is 0xff_ff_ff_ff, this should be more than enough for tabled
+///   symbols.
+///
+/// For interners other than the one in `SESSION_GLOBALS` (which are rare and
+/// not that performance-critical) all symbols are interned, i.e. the u32 is an
+/// index into the interner's table with its high bit set.
 #[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct Symbol(SymbolIndex);
 
 rustc_index::newtype_index! {
     pub struct SymbolIndex { .. }
 }
 
+// These `Symbol` methods are simple methods that work with just the `Symbol`
+// itself, and do not involve `SESSION_GLOBALS`.
 impl Symbol {
-    const fn new(n: u32) -> Self {
-        Symbol(SymbolIndex::from_u32(n))
+    /// Try to create an inlined symbol from a string.
+    /// WARNING: this function must behave equivalently to `is_inlinable()` in
```

Comment:
It's not exactly the same. `is_inlinable` checks that all chars are not zero and `< 0x80`, this function only checks that for the last char in each string. So either the docs need to state that `is_inlinable` must be stricter or equal to this function, or this function needs to be fixed.

Reply:
Maybe some well placed debug assertions could ensure the two functions are in sync

Reply:
I tried to make this clear by the use of "equivalently" and the comment on `is_inlinable()`, but I guess I failed. I will try to clarify the comment further.